auth/api/iam/openid4vp.go

```diff
@@ -286,6 +309,253 @@ func (r Wrapper) sendAndHandleDirectPostError(ctx context.Context, auth2Error oa
 	}, nil
 }
 
+func (r Wrapper) HandleAuthorizeResponse(ctx context.Context, request HandleAuthorizeResponseRequestObject) (HandleAuthorizeResponseResponseObject, error) {
+	// this can be an error post or a submission. We check for the presence of the error parameter.
+	if request.Body.Error != nil {
+		return r.handleAuthorizeResponseError(ctx, request)
+	}
+
+	// successful response
+	return r.handleAuthorizeResponseSubmission(ctx, request)
+}
+
+func (r Wrapper) handleAuthorizeResponseError(_ context.Context, request HandleAuthorizeResponseRequestObject) (HandleAuthorizeResponseResponseObject, error) {
+	// we know error is not empty
+	code := *request.Body.Error
+	var description string
+	if request.Body.ErrorDescription != nil {
+		description = *request.Body.ErrorDescription
+	}
+
+	// check if the state param is present and if we have a client state for it
+	var oauthSession OAuthSession
+	if request.Body.State != nil {
+		if err := r.oauthClientStateStore().Get(*request.Body.State, &oauthSession); err == nil {
+			// we use the redirectURI from the oauthSession to redirect the user back to its own error page
+			if oauthSession.redirectURI() != nil {
+				location := httpNuts.AddQueryParams(*oauthSession.redirectURI(), map[string]string{
+					oauth.ErrorParam:            code,
+					oauth.ErrorDescriptionParam: description,
+				})
+				return HandleAuthorizeResponse200JSONResponse{
+					RedirectURI: location.String(),
+				}, nil
+			}
+		}
+	}
+	// we don't have a client state, so we can't redirect to the holder redirectURI
+	// return an error page instead
+	return nil, oauthError(oauth.ErrorCode(code), description)
+}
+
+func (r Wrapper) handleAuthorizeResponseSubmission(ctx context.Context, request HandleAuthorizeResponseRequestObject) (HandleAuthorizeResponseResponseObject, error) {
+	verifier, err := r.idToOwnedDID(ctx, request.Id)
+	if err != nil {
+		return nil, oauthError(oauth.InvalidRequest, "unknown verifier id")
+	}
+
+	if request.Body.VpToken == nil {
+		return nil, oauthError(oauth.InvalidRequest, "missing vp_token")
+	}
+
+	pexEnvelope, err := pe.ParseEnvelope([]byte(*request.Body.VpToken))
+	if err != nil || len(pexEnvelope.Presentations) == 0 {
+		return nil, oauthError(oauth.InvalidRequest, "invalid vp_token")
+	}
+
+	// note: instead of using the challenge to lookup the oauth session, we could also add a client state from the verifier.
+	// this would allow us to lookup the redirectURI without checking the VP first.
+
+	// extract the nonce from the vp(s)
+	nonce, err := extractChallenge(pexEnvelope.Presentations[0])
```

Comment:
should check all presentations. this will have 2 VPs (user and care org)

Reply:
this is done 20 lines later